### PR TITLE
refactor: rework merge action structure

### PR DESCRIPTION
**Pull request by sileht**

### CODE REVIEW

File: mergify_engine/actions/__init__.py [hunk 1]
```diff
@@ -89,12 +89,8 @@ def command_to_config(string):  # pragma: no cover
         """Convert string to dict config"""
         return {}
 
-    def run(
-        self, ctxt, rule, missing_conditions
-    ) -> check_api.Result:  # pragma: no cover
+    def run(self, ctxt, rule) -> check_api.Result:  # pragma: no cover
```
Comment (jd) [inline mergify_engine/actions/__init__.py:92]: more typing?

File: mergify_engine/actions/__init__.py [hunk 2]
```diff
@@ -89,12 +89,8 @@ def command_to_config(string):  # pragma: no cover
...
         pass
 
-    def cancel(
-        self, ctxt, rule, missing_conditions
-    ) -> check_api.Result:  # pragma: no cover
+    def cancel(self, ctxt, rule) -> check_api.Result:  # pragma: no cover
```
Comment (jd) [inline mergify_engine/actions/__init__.py:95]: ditto

File: mergify_engine/actions/assign.py [hunk 3]
```diff
@@ -32,7 +32,7 @@ class AssignAction(actions.Action):
 
     silent_report = True
 
-    def run(self, ctxt, rule, missing_conditions) -> check_api.Result:
+    def run(self, ctxt, rule) -> check_api.Result:
```
Comment (jd) [inline mergify_engine/actions/assign.py:35]: ditto

File: mergify_engine/actions/backport.py [hunk 4]
```diff
@@ -30,7 +30,7 @@ class BackportAction(copy.CopyAction):
     def command_to_config(string):
         return {"branches": string.split(" ")}
 
-    def run(self, ctxt, rule, missing_conditions) -> check_api.Result:
+    def run(self, ctxt, rule) -> check_api.Result:
```
Comment (jd) [inline mergify_engine/actions/backport.py:33]: ditto

File: mergify_engine/actions/close.py [hunk 5]
```diff
@@ -30,7 +30,7 @@ class CloseAction(actions.Action):
     only_once = True
     validator = {voluptuous.Required("message", default=MSG): types.Jinja2}
 
-    def run(self, ctxt, rule, missing_conditions) -> check_api.Result:
+    def run(self, ctxt, rule) -> check_api.Result:
```
Comment (jd) [inline mergify_engine/actions/close.py:33]: ditto

File: mergify_engine/actions/comment.py [hunk 6]
```diff
@@ -28,7 +28,7 @@ class CommentAction(actions.Action):
 
     silent_report = True
 
-    def run(self, ctxt, rule, missing_conditions) -> check_api.Result:
+    def run(self, ctxt, rule) -> check_api.Result:
```
Comment (jd) [inline mergify_engine/actions/comment.py:31]: ditto

File: mergify_engine/actions/copy.py [hunk 7]
```diff
@@ -104,7 +104,7 @@ def _copy(self, ctxt, branch_name):
             f"{self.KIND.capitalize()} to branch `{branch_name}` failed",
         )
 
-    def run(self, ctxt, rule, missing_conditions) -> check_api.Result:
+    def run(self, ctxt, rule) -> check_api.Result:
```
Comment (jd) [inline mergify_engine/actions/copy.py:107]: ditto

File: mergify_engine/actions/delete_head_branch.py [hunk 8]
```diff
@@ -29,7 +29,7 @@ class DeleteHeadBranchAction(actions.Action):
         {voluptuous.Optional("force", default=False): bool}, None
     )
 
-    def run(self, ctxt, rule, missing_conditions) -> check_api.Result:
+    def run(self, ctxt, rule) -> check_api.Result:
```
Comment (jd) [inline mergify_engine/actions/delete_head_branch.py:32]: i'm tired of writing dito now.

File: mergify_engine/actions/merge.py [hunk 9]
```diff
@@ -243,7 +247,7 @@ def _required_statuses_in_progress(ctxt, missing_conditions):
 
         return False
 
-    def _sync_with_base_branch(self, ctxt, rule, missing_conditions, q):
+    def _sync_with_base_branch(self, ctxt, rule, q):
```
Comment (jd) [inline mergify_engine/actions/merge.py:250]: typing?

File: mergify_engine/actions/merge.py [hunk 10]
```diff
@@ -479,3 +475,158 @@ def _handle_merge_error(
                 message,
                 f"GitHub error message: `{e.message}`",
             )
+
+    @staticmethod
+    def merge_report(ctxt, strict) -> typing.Union[check_api.Result, None]:
```
Comment (jd) [inline mergify_engine/actions/merge.py:480]:
```suggestion
    def merge_report(ctxt, strict) -> typing.Optional[check_api.Result]:
```

File: mergify_engine/actions/merge.py [hunk 11]
```diff
@@ -479,3 +475,158 @@ def _handle_merge_error(
...
+
+        # NOTE(sileht): remaining state "behind, clean, unstable, has_hooks
+        # are OK for us
+        else:
+            return None
+
+        return check_api.Result(conclusion, title, summary)
+
+    @staticmethod
+    def get_queue_summary(ctxt):
```
Comment (jd) [inline mergify_engine/actions/merge.py:543]: typing

**Review by jd — APPROVED** [review]

**Comment by sileht** [comment]
Eheh I have done all the typing before I read the review :)

**Pull request merged**